### ISSUE TITLE
Support serializing executable suite into JSON

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#3902]
It would be convenient to be able to serialize executable `TestSuite` objects into JSON that can be then also be used to recreate same suite later. Most important use cases are:

- Easier to transfer tests between processes and machines. This would help, for example, https://github.com/MarketSquare/robotframework-cluster.
- Possibility to save a suite, possible a nested suite, constructed from data on file system into a single file that is faster to parse.
- Alternative data format for external tools generating tests.

In the subsequent comments I'll go through some design ideas I have related to this. Please use :+1: and :-1: to indicate do you like those ideas as well as this issue in general. Written comments obviously appreciated as well!

**Comment by pekkaklarck** [comment]
I propose we add `to_json` method to `TestSuite`. It should by default return JSON string but it should also accept a path or an open file object where to save the JSON. It could also accept an option to specify should the object be serialized into a single JSON object or split to multiple. I write a separate comment about that.

We should also add `from_json` classmethod to `TestSuite` for constructing a suite based on JSON data.

We probably should add `to_json` and `from_json` also to `TestCase`, `Keyword`, `If` and `For` objects.

**Comment by pekkaklarck** [comment]
Serializing a big suite into JSON is problematic because reading it back requires deserializing the big JSON string first. To avoid problems with that, I propose we support splitting suites into multiple objects using this algorithm:

- The top level suite is one object.
- Suites, tests, setup and teardown a suite has are separate objects. Each separate object has a growing index. The parent suite has this index in the place of the real object.
- All objects are stored into a single file, one object per line. This is not valid JSON but streaming JSON like this is pretty common. The format is described as two pretty identical standards [JSON Lines](https://jsonlines.org) and [Newline Delimited JSON](http://ndjson.org/).
- Child object index is got directly from the order they are serialized. The first child has index 1, the second index 2, and so on.

I propose we support both "one big object" and "split object" formats when construction model objects using `from_json`. The `to_json` should get an optional `split` argument. I think it should be `False` by default to produce valid JSON unless explicitly configured.

An example of a suite as one object:

    {"name": "Suite", "doc": "Example", "tests": [{"name": "T1", "body": [{"name": "Log", "args": ["Hi!"]}]}, {"name": "T2"}]}

Same suite split to multiple objects:

    {"name": "Suite", "doc": "Example", "tests": [1, 2]}
    {"name": "T1", "body": [3]}
    {"name": "T2"}
    {"name": "Log", "args": ["Hi!"]}

Above examples use spaces between items but I believe we should omit them to save some space.

**Comment by pekkaklarck** [comment]
I propose we use new `.rbt` extension when saving this format into a file. We cannot really use `.json` if we support also JSON Lines and having a custom extension also separates these files from other JSON files. We cannot use `.robot` because the format is different but we can consider `.rbt` to mean "condensed robot".

We should allow running `robot suite.rbt` as well as `robot directory` where the directory contains these files.

**Comment by pekkaklarck** [comment]
Saved suite objects in `.rbt` files should also contain imports, variables and keywords same way as normal `.robot` files can have.

**Comment by pekkaklarck** [comment]
External resource files could possibly be saved into a `.rbt` file as well. The best approach would be including all imported resources in it automatically but this probably should be configurable. The biggest problem would be designing how to change imports so that original `example.resource` would point to an embedded resource. This requires changes to the import machinery and probably needs a separate issue. Could also be implemented later.

**Comment by pekkaklarck** [comment]
This issue is highly related to the proposal to add JSON based output format (#3423). Executable suite structure and result suite structure are similar, they even have common base, and they should use the same approach to serialize them to JSON. I think splitting a suite into multiple objects like I [propose above](https://github.com/robotframework/robotframework/issues/3902#issuecomment-805280596) ought to work fine also with result suites. They couldn't use the [proposed .rbt extension](https://github.com/robotframework/robotframework/issues/3902#issuecomment-805289194), though, because they contain different data and we don't want `robot directory` to try running possible result files.

**Comment by pekkaklarck** [comment]
The biggest reason to split JSON into multiple objects is to avoid the need to read the whole file into memory and parse the whole JSON in one go. This may not be a huge problem with data, but if this same approach is used with results (#3423) it's very important because I've seen output.xml files that are over 1GB.

Assuming we have this JSON from the [earlier example](https://github.com/robotframework/robotframework/issues/3902#issuecomment-805280596)
```
{"name": "Suite", "doc": "Example", "tests": [1, 2]}
{"name": "T1", "body": [3]}
{"name": "T2"}
{"type": "KEYWORD", "name": "Log", "args": ["Hi!"]}
```
it could be processed like this:

1. Open the file and iterate over it like `for line in file` to read only one line at a time.
2. Process the first line. Construct `TestSuite` object with specified `name` and `doc`. Tests are not specified directly (there are integers, not objects) so add callbacks to add them when child objects with id 1 and 2 are encountered.
3. Process the second line. This line has implicit id 1 and thus a callback created earlier is called with this data creating a `TestCase` object that's added to the earlier suite. Tests body has one item referred with id 3. This creates another callback.
4. Process the second line (id 2). Call relevant callback to created another `TestCases`.
5. Process the last line (id 3). Call relevant callback to created `Keyword` that's added to the first test.

**Comment by pekkaklarck** [comment]
Too big task for RF 4.1. Hopefully can be included in RF 5.0.

**Comment by awasekhirni** [comment]
Wishing list as a robotframework/RPA development 
1. Ability to save TestSuite files (dynamically generated as) as '.robot' and '.json' for reuse.
2. robot.api documentation and crud operations examples would help many people completely automate the automation process 
3. Ability to save the objecthierarchy prior to execution as a json file with all the resources, testcases, pageobjects,keywords, webelementlocators as a JSON file

**Comment by awasekhirni** [comment]
Another item i missed out is 
1. Ability to write the ".robot" file as a json file. When RIDE supports "double tabbed" and "pipe delimiter" format for '.robot' file creation. Why not allow the user to create'.robot' file in json format and allow him to execute the same. this is industry wide standard and would prove beneficial to all making automation of automation more easier

**Comment by awasekhirni** [comment]
from robot.api.parsing import (
    get_model, Documentation, EmptyLine, KeywordCall,
    ModelTransformer, SettingSection, SectionHeader, Token,Documentation,Tags,DefaultTags,ForceTags,
    Setup,SuiteSetup,TestSetup,EmptyLine,Teardown
)

# Awase Khirni Syed awase008@gmail.com

class CaseModifier(ModelTransformer):

    def visit_TestCase(self, node):
        # The matched `TestCase` node is a block with `header` and
        # `body` attributes. `header` is a statement with familiar
        # `get_token` and `get_value` methods for getting certain
        # tokens or their value.
        name = node.header.get_value(Token.TESTCASE_NAME)
        # Returning `None` drops the node altogether i.e. removes
        # this test.
        for i,tcase in enumerate(lists):
            if name == 'loginTest':
                node.body.append(Documentation.from_params("This is my documentations to be added dynamically"))
                node.body.append(Tags.from_params("Mydynamic Tag goes here"))
                node.body.append(Setup.from_params("Mydynamic setup goes here"))
                new_keyword = KeywordCall([
                    Token(Token.SEPARATOR, '    '),
                    Token(Token.KEYWORD, 'New Keyword'),
                    Token(Token.SEPARATOR, '    '),
                    Token(Token.ARGUMENT, 'xxx'),
                    Token(Token.EOL)
                ])
                node.body.append(new_keyword)
                node.body.append(Teardown.from_params("Mydynamic teardown"))
                node.body.append(EmptyLine.from_params())

                #node.body.insert(i, new_keyword)

            return node


        # Construct new keyword call statement from tokens. See `visit_File`
        # below for an example creating statements using `from_params`.

        # Add the keyword call to test as the second item.

        # No need to call `generic_visit` because we are not
        # modifying child nodes. The node itself must to be
        # returned to avoid dropping it.
        return node

    def visit_File(self, node):

        #Removing the old settings
        #node.sections.pop(0)
        # Create settings section with documentation. Needed header and body
        # statements are created using `from_params` method. This is typically
        # more convenient than creating statements based on tokens like above.
        settings = SettingSection(
            header=SectionHeader.from_params(Token.SETTING_HEADER),
            body=[
                Documentation.from_params('This is a really\npowerful API!'),
                EmptyLine.from_params()
            ]
        )
        # Add settings to the beginning of the file.
        node.sections.insert(0, settings)

        # vset=SettingSection(header=SectionHeader.from_params(Token.VARIABLE_HEADER),body="Variables")
        # node.sections.insert(1,vset)
        #tset = SettingSection(header=SectionHeader.from_params(Token.TESTCASE_HEADER), body="Variables")
        # node.sections.insert(2,tset)
        #keyset=SettingSection(header=SectionHeader.from_params(Token.KEYWORD_HEADER),body='Keywords')
        #node.sections.insert(3,keyset)

        # Call `generic_visit` to visit also child nodes.
        return self.generic_visit(node)


    def visit_Tag(self,node):
            node.body.insert(0,'smoketest')
            return node


model = get_model('example.robot')
TestModifier().visit(model)
model.save('modified.robot')

**Comment by pekkaklarck** [comment]
@mkorpela, this issue is about saving the running side model to JSON that could then be later easily parsed by Robot and executed. There's separate #3423 about output.json i.e. results in JSON format.

**Comment by pekkaklarck** [comment]
@awasekhirni, this issue covers most of the features you want. Somehow including also referenced resource and variable files or even libraries is out of its scope. Once this issue is done, we can look at those enhancements as well.

**Comment by pekkaklarck** [comment]
I've been lately thinking this issue a bit because we could possibly use the same underlying logic to convert model objects to dictionaries (that can then later trivially be converted to JSON) could possibly also used elsewhere. For legacy reasons there's still code that considers everything tests contains keywords although there can be FOR loops, IF/ELSE structures, and after RF 5.0 also RETURN, BREAK, CONTINUE, and TRY/EXCEPT objects. The current non-keyword objects have attributes that allow using them as keywords, but that requires quite a bit of code here and there and is ugly in general. I want this compatibility code to be removed sooner rather than later, but there still should be a way to somehow uniformly operate all "body items" tests can contain.

A solution I've been thinking is adding method `to_dict(as_keyword=False)` to all body items. By default they'd return all their data as a dictionary with structure unique to that particular object type. If `as_keyword=True` would be used, they'd return information as a dictionary containing same keys as keyword objects have. This would allow working with all these objects "as keywords" if needed.

If the aforementioned `to_dict` is implemented to all body items tests can have, we could easily add `to_dict` also to tests and suites. Then it would be trivial to add `to_json` to all these that would just call `to_dict` and convert the result to JSON. That would mean half of this issue would be done. The other half would be adding `from_dict` and `from_json` class methods for constructing model objects.

**Comment by pekkaklarck** [comment]
At the moment it seems unlikely this issue will make it to RF 5.0. There are other more important issues and when they are done we probably want to create the release so that we get them into real use.

**Comment by manykarim** [comment]
Really looking forward to that feature.
I'm playing around with AWS Lambda at the moment.
And that feature would (hopefully) allow me, to trigger a prepared Lambda Function (for test execution) and only transfer the test cases to be executed as json.
Right now, I need to transfer the complete testsuite and run some prerunmodifier to split things up.
Not sure if that makes sense - but can't wait to try it out.

**Comment by pekkaklarck** [comment]
Lambda usage sounds like a very good use case for this functionality. This is currently targeted for RF 6.0, but we could still look at including it into RF 5.2. Are you @manykarim interested to help with the implementation? The first step would deciding the format to use. We can discuss that here or start a new thread on #devel on our Slack.

**Comment by pekkaklarck** [comment]
I read  through my earlier comments about the possible design and the proposed approach still looks pretty good. The main open design decision seems to be deciding how to store imports, variables and keywords into `*.rbt` files as well as deciding should we include resource file contents into them.

**Comment by pekkaklarck** [comment]
This would be a really useful functionality and I hope we could get it implemented already in RF 6.1, not in 7.0. I'll set milestone and priority accordingly.

**Comment by pekkaklarck** [comment]
This is the most important RF 6.1 issue because there are so many interesting usages for it in the ecosystem. Basic functionality is ready and will be part of the forthcoming RF 6.1 alpha 1. This is what works:

1. You can serialize a suite structure into JSON by using [TestSuite.to_json](https://robot-framework.readthedocs.io/en/latest/autodoc/robot.running.html#robot.running.model.TestSuite.to_json) method. When used without arguments, it returns the JSON as a string, but it also accepts a path or an open file where to write it along with configuration options related to JSON formatting. Example:
    ```python
    from robot.api import TestSuite
  
    suite = TestSuite.from_file_system('path/to/tests')
    suite.to_json('tests.rbt')
    ```

2. You can create a suite based on JSON data using [TestSuite.from_json](https://robot-framework.readthedocs.io/en/latest/autodoc/robot.running.html#robot.running.model.TestSuite.from_json). It works both with JSON strings and paths to JSON files. Example:
    ```python
    from robot.api import TestSuite
  
    suite = TestSuite.from_json('tests.rbt')
    ```

3. When using `robot` normally, it parses `.rbt` files automatically. This includes running individual JSON files like `robot tests.rbt` and running a directory containing them.

**Comment by pekkaklarck** [comment]
Remaining tasks to get this done:
- [ ] Handling resource files.
   - [ ] Support serializing resource files to JSON and using them with "normal" `.robot` tests. This requires deciding on a suitable JSON resource file extension. `*.rsrc` perhaps?
   - [ ] Support including resource files into the suite JSON. This requires deciding should we do it always or should it be opt-in. This may also turn out to be too complicated to do in RF 6.1 timeframe.
- [ ] Write documentation. API docs ought to be already pretty good, but this needs to be documented in the User Guide as well.
- [ ] Test execution of `.rbt` files.
- [ ] Check are tests for serialization and deserialization adequate.
- [ ] Decide do we need external entry point for creating `.rbt` files. I believe it's not needed at this point, but if others think it's a good idea and have an idea how it should work, this certainly can still be considered.

Anything else that's still missing? Any opinions about handling resource files? Ping especially @manykarim.

**Comment by luv-deluxe** [comment]
I know it might be a bit late now, but perhaps once model is loaded from JSON (*.rbt), one could update its variable with e.g.

```python
model = TestSuite().from_json()
model.update_variable(var="myvar", value=15)
model.update_variable(var="listvar", value=["Bill", 0, None])

# it updates model body, but one can imagine that saved ROBOT would look like below
# ***SETTINGS***
# ${myvar}    ${15}
# @{names}    Bill    ${0}    ${None}
```

As I remember, variables are stored as `Variable()` objects having `value` attribute that returns a tuple.
It would be helpful for `robotframework-cluster` or other distributed execution. Serialized ROBOT is sent to a worker, that updates variables read from other place (database, environment or configmap) and executes it directly. No need for variable assignment via CLI or variables file.

**Comment by pekkaklarck** [comment]
You can manipulate `TestSuite.resource.variables` but it's not too convenient. That API could be enhanced, but I believe you typically want to pass variables like this to `TestSuite.run` and that works already now.

**Comment by pekkaklarck** [comment]
The above commit added support for JSON resource files and enhanced error handling if a parsed JSON file is invalid. With resource files it is possible to use either `*.json` or `*.rsrc` extension.

**Comment by pekkaklarck** [comment]
There are currently the following limitations:

1. The source path is stored in absolute format. This causes problems if data is moved to another machine with a different directory layout. The main problem is that imports that use relative paths are relative to the source. This is a pretty bad limitation, but I'm not sure what's the best way to handle it. Possibly `to_dict/json` could allow making the source relative to a certain directory, and `from_dict/json` could then turn the relative path to an absolute path based on where the source file is located. This could possibly still be fixed in RF 6.1 but I don't consider it mandatory.
2. Resource and variable files imported by a suite cannot be bundled with it. Fixing this would require making it possible to resolve imports without executing tests/tasks. Too much work for RF 6.1.
3. `${CURDIR}` doesn't work. Fixing it would basically require fixing #596 and that's too much work for RF 6.1.
4. Documentation is missing. This obviously must be fixed before RF 6.1.

The limitations listed above above need to mentioned in the documentation.

**Comment by pekkaklarck** [comment]
In the previous comment I mentioned that absolute suite source is problematic. The commit above added `TestSuite.adjust_source` that can be used to make absolute source relative and to add a new root to a relative source. That's especially convenient when moving JSON data around. This example is from the docstring of the new method:
```python
from robot.running import TestSuite                      
                                                         
# Create a suite, adjust source and convert to JSON.     
suite = TestSuite.from_file_system('/path/to/data')      
suite.adjust_source(relative_to='/path/to')              
suite.to_json('data.json')                               
                                                         
# Recreate suite elsewhere and adjust source accordingly.
suite = TestSuite.from_json('data.json')                 
suite.adjust_source(root='/new/path/to')                 
```

After this enhancement, I consider JSON serialization support to be good enough for RF 6.1. Documentation is still missing, but I'll write it after the first release candidate is out.

**Comment by pekkaklarck** [comment]
Commits listed above have added a schema as well as documentation in the User Guide. Creating JSON resource files was made more convenient by #4793. We can finally consider this issue done.
